src/server/pkg/storage/chunk/writer.go

```diff
@@ -30,12 +30,22 @@ var initialWindow = make([]byte, WindowSize)
 // WriterFunc is a callback that returns a data reference to the next chunk and the annotations within the chunk.
 type WriterFunc func(*DataRef, []*Annotation) error
 
+// byteSet is a unit of work for the workers.
+// A worker will roll the rolling hash function across the bytes
+// while processing the associated annotations.
 type byteSet struct {
 	data        []byte
 	annotations []*Annotation
-	nextByteSet chan *byteSet
+	// nextBytes is used for the edge case where no split point is found in an assigned byte set.
+	// A worker that did not find a chunk in its assigned byte set will pass it to the prior
+	// worker in the chain with nextBytes set to the next worker's bytes channel. This allows shuffling
+	// of byte sets between workers until a split point is found.
```

Comment:
It seems like this daisy chaining process adds a lot of complexity to the process and I don't understand why it's necessary. I think this comes from the fact that if you split the bytes up among the workers and they don't find split points it's still possible that there's a split point but the window which satisfies the split predicate is spread between 2 different workers. Is that the reason for the shuffle?

Reply:
Daisy chaining is actually a pretty simple model for serializing the results of the processing while also allowing parallel processing. It trades a master that coordinates and shuffles edge byte sets for workers knowing the prior/next worker in the chain and communicating with them. I am all ears if you have an even simpler model in mind, but I think the daisy chain model is very simple but could maybe be refactored a bit more. Just to be clear, the daisy chaining is necessary for the serialization of results (executing the callback in the order that the chunks appear in the byte stream), not this particular edge case (although it is handled by piggybacking on the daisy chain).

Reply:
It sounds like one source of complexity is that the splitting process and the calling of callbacks are coupled together and one of them wants to parallel and the other needs to be serial. Does it particularly matter that we call the callbacks quickly? Or could we decouple them from the splitting process? The split window edge case also seems like it can be solved by including a window's worth of overlapping bytes so that every window can be evaluated by exactly one worker. I'm imagining this would you allow the code to look a little bit more like a standard split and merge concurrency pattern:

```
var eg errgroup.Group
splits := make([][]int, len(chunks))
for i, chunk := range chunks {
  eg.Go(func() error {
    splits[i] = computeSplits(chunk)
  }
}
eg.Wait()
for _, s := range splits {
  // call cbs
}
```

It seems to me like the need for the daisy-chain pattern is coming from corner-cases that aren't inherent to the problem and can be avoided if you architect things slightly differently. In which case I think it's worth it to tweak things a bit so we can use something simpler. I could be missing something though.

Reply:
The algorithm you provided, as is, does not support streaming or workloads that do not fit in memory. You would need some sort of batching logic that would also resolve splits/annotations across the batches. The basic pattern would become master/workers where the master would be shuffling information from/to workers and executing callbacks as compared to setting up a chain of channels (`chanSet`). Also, I think you might be overlooking the cost of hashing the chunks and annotations that overlap with the model you proposed. Are the overlapping chunks/annotations just going to be hashed by the master and stuck into the right place in terms of the callbacks? That would be some extra complexity and inefficient. If not, then the information about chunks/annotations that hang over the splits will need to be shuffled between workers by the master which is a non-trivial amount of complexity. The nice thing about daisy chaining is that everything is parallelized and there is no middleman for shuffling. It is necessary to have some information shuffled between workers to fully parallelize the work. I think the key difference is should a chain be setup for shuffling and serializing results or should there be a master shuffling data and serializing the results. I still think the chain is simpler.

Reply:
Alright, I think I can see how this operation fits better into a daisy-chained model than a master/worker model. One last thing that I think would improve this code is I think you can probably add some directional limitations to the channel types. Which I think will make the code a bit more understandable, and also allow the compiler to catch a few more errors.

Reply:
Done.